### PR TITLE
Add comprehensive matrix library with advanced algorithms

### PR DESCRIPTION
**Pull request by AkshaySG14**
## Summary

This PR introduces a complete matrix computation library with extensive functionality for linear algebra, geometric transformations, and numerical algorithms.

### Core Features
- **Matrix Class**: Full-featured matrix implementation with operator overloading
  - Addition, subtraction, multiplication (matrix and scalar)
  - Transpose, equality checking, indexing
  - Factory methods: `identity()`, `zeros()`, `ones()`
  - Row/column access methods

### Operations Module
- Basic operations: add, subtract, multiply
- Vector operations: dot product
- Matrix properties: trace, determinant, various norms
- Efficient implementations for common operations

### Transformations Module  
- **2D Transformations**:
  - Rotation, scaling, shearing
  - Reflection (x-axis, y-axis, origin)
- **3D Transformations**:
  - Rotation about x, y, z axes
  - Translation with homogeneous coordinates
- Composable transformation matrices

### Advanced Algorithms
- **Gaussian Elimination**: Row echelon form computation
- **LU Decomposition**: Efficient for solving multiple systems
- **QR Decomposition**: Gram-Schmidt orthogonalization
- **Linear System Solver**: Solve Ax = b with back substitution
- **Rank Computation**: Matrix rank via row reduction

### Testing & Documentation
- 30+ comprehensive unit tests covering all modules
- Integration tests for complex transformation chains
- Full API documentation with usage examples
- Interactive demo application showcasing features

## Files Changed
- `matrix_lib/` - Core library modules (5 files)
- `tests/` - Test suite  
- `docs/API.md` - Complete API documentation
- `main.py` - Interactive demo application

## Test Plan
- [x] Matrix class operations tested (10+ tests)
- [x] All transformation functions verified
- [x] Algorithm correctness validated
- [x] Edge cases and error handling covered
- [x] Integration tests for composed operations

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: matrix_lib/matrix.py [hunk 1]
```diff
@@ -0,0 +1,183 @@
+"""Core Matrix class implementation."""
+
+from typing import List, Union, Tuple, Optional
+import math
+
+
+class Matrix:
+    """
+    A flexible matrix class with comprehensive operations.
```
Comment (AkshaySG14) [inline matrix_lib/matrix.py]: Saar?
